src/ir/ExpressionAnalyzer.cpp

```diff
@@ -483,27 +280,47 @@ size_t ExpressionAnalyzer::hash(Expression* curr) {
         // call_imports type, etc. The simplest thing is just to hash the
         // type for all of them.
         rehash(digest, curr->type.getID());
-        // Blocks and loops introduce scoping.
-        if (auto* block = curr->dynCast<Block>()) {
-          noteScopeName(block->name);
-        } else if (auto* loop = curr->dynCast<Loop>()) {
-          noteScopeName(loop->name);
-        } else {
-          // For all other nodes, compare their immediate values
-          visitImmediates(curr, *this);
```

Comment:
I think this PR also reduces the time complexity from O(n^2) to O(n) because before when we visit every child of a node for every node visited by calling `visitImmediates` for ever node, but now we only visit each node once.. Is this correct?

Reply:
Not sure I follow. What is `n`?

I think the complexity is similar. Before we would call `visitImmediates` which would accumulate the immediates to vectors. That's `O(number of immediates)` (times the number of nodes). We do the same now, but we avoid needing the vectors.

Reply:
For example, if there's an expression
```wasm
(i32.add
  (i32.sub
    (i32.const 3)
    (i32.const 4)
  )
)
```
In this case, wasn't `visitImmediates` called for every single node (`add`, `sub`, two `const`s) here? And `visitImmediates` ran a walker that traverses all its children. As a result, the leaf nodes were visited not only once but every time any of their parent nodes was visited..? Now we use a stack, each node seems to be only visited once. May be I'm confused about something.

Reply:
Before, each call to `visitImmediates` would visit the immediates of that node, but not of child nodes. `visitImmediates` is a visitor, not a walker.

Reply:
(maybe we need better terms for those, but I don't have an idea myself...)

Reply:
Ah right I was confused... Thanks for the explanation.